scipy/optimize/_optimize.py

```diff
@@ -131,6 +132,45 @@ def _dict_formatter(d, n=0, mplus=1, sorter=None):
     return s
 
 
+def _wrap_callback(callback):
+    """Wrap a user-provided callback so that attributes can be attached."""
+    params = (inspect.Parameter('res_i', kind=inspect.Parameter.KEYWORD_ONLY),)
```

Comment:
why keyword only? Can't we search for intermediate_result anywhere in the signature?

Reply:
We could. I that if we say keyword-only, the user is less likely to assume that it can be called whatever they want. It's really just for emphasis that it has to be called `intermediate_result`.